htlcswitch/queue.go

```diff
@@ -152,9 +152,18 @@ func (p *packetQueue) packetCoordinator() {
 			case p.outgoingPkts <- nextPkt:
 				// Only decrease the queueLen and totalHtlcAmt once the packet has been
 				// sent out
-				debug_print("going to remove nextPkt from the queue")
+				// FIXME: do we need these to be atomic? Since the queue is per channel
+				// link, I think only one thread can be updating this at a time?
 				atomic.AddInt32(&p.queueLen, -1)
 				atomic.AddInt64(&p.totalHtlcAmt, int64(-nextPkt.amount))
+				// update the minHtlc value
```

Comment:
@vibhaa 	yeah this is a good point. I had thought that the atomics were added when we weren't yet sure about whether the queue serves a single channel v/s multiple channels etc. but they seem to be there from the original lnd code base. Presumably, a switch could get multiple payments over a channel, which may be serviced in separate goroutines. A possible explanation was in the comments:

// queueLen is an internal counter that reflects the size of the queue
// at any given instance. This value is intended to be use atomically
// as this value is used by internal methods to obtain the length of
// the queue w/o grabbing the main lock. This allows callers to avoid a
// deadlock situation where the main goroutine is attempting a send
// with the lock held.

In link.go, they have used queue.Length() function to check the length of the queue when processing a new packet - which is presumably what the comment above was referring to. For the minHtlcAmt variable: we update its value directly only in queue.go --> when adding and dequeuing a packet. This can be protected by the main queue lock (in AddPkt, the lock was already being held, and I also added it now to the dequeue situation above).

But we also access its value in the periodic goroutine in link.go which decides whether to signal to the queue or not. In general though, it getting stale values isn't really a problem - even if it signals the queue spuriously, it only causes the queue to attempt to send a packet which possibly couldn't be send (something that can happen anyway since we are serving from the front of the queue and not the minimum...). Thus I suppose this is not a problem right now, but should be something to keep in mind (?)

we could also put the access from link.go under the same main queue lock - but i am not sure if it is getting a bit excessive locking there -- also maybe we will not need the minHtlcAmt variable anyway if we change our dequeuing strategy...